File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] X = new int[N];
        boolean[] exist = new boolean[1000001];
        int maxX = 0;
        for (int i = 0; i < N; i++) {
            X[i] = Integer.parseInt(st.nextToken());
            exist[X[i]] = true;
            if (maxX < X[i])
                maxX = X[i];
        }
        int[] score = new int[maxX + 1];
        for (int x : X) {
            for (int i = x + x; i <= maxX; i += x) {
                if (exist[i]) {
                    score[i]--;
                    score[x]++;
                }
            }
        }
        for (int x : X)
            sb.append(score[x]).append(" ");
        System.out.println(sb);
    }

}

File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.py
import sys
input = sys.stdin.readline

def solve():
    N = int(input())
    X = [*map(int, input().split())]
    exist = [0] * 1000001
    maxX = 0
    for x in X:
        exist[x] = 1
        if maxX < x:
            maxX = x
    score = [0] * (maxX + 1)
    for x in X:
        for v in range(x + x, maxX + 1, x):
            if exist[v]:
                score[v] -= 1
                score[x] += 1
    for x in X:
        print(score[x], end=" ")

solve()